Geant4e integration in 7_4_X

lviliani opened the pull request:
Added TrackPropagation/Geant4e package.
Some changes to RecoTracker/TrackProducer, TrackingTools/PatternTools to enable the propagation to PCA using a non analytical propagator.
Changes to SimG4Core/GeometryProducer in order to allow the propagation with Geant4e.

lviliani committed:
Geant4e

lviliani committed:
Geant4e

lviliani committed:
Delete geant4.10.p2_exitNormal.patch

lviliani committed:
Cleanup

lviliani committed:
Cleanup

lviliani committed:
Cleanup

lviliani committed:
Cleanup

cmsbuild commented:
A new Pull Request was created by @lviliani (Lorenzo Viliani) for CMSSW_7_4_X.

Geant4e integration in 7_4_X

It involves the following packages:

RecoTracker/TrackProducer
SimG4Core/GeometryProducer
TrackPropagation/Geant4e
TrackingTools/PatternTools


@civanch, @nclopezo, @cvuosalo, @mdhildreth, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@ghellwig, @battibass, @makortel, @abbiendi, @GiacomoSguazzoni, @rovere, @VinInn, @appeltel, @bellan, @gpetruc, @cerati, @trocino, @dgulhan, @venturia this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

civanch commented:
@lviliani, in SimG4Core/GeometryProducer/interface/GeometryProducer.h you really need sensitive detectors in some cases?

cmsbuild commented:
+1
Tested at: dc877f797b555f1b83e0995753ecb8064aa62b46
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7505/2297/summary.html

davidlange6 commented on TrackingTools/PatternTools/BuildFile.xml:
```diff
@@ -12,6 +12,7 @@
 <use   name="TrackingTools/TrajectoryParametrization"/>
 <use   name="TrackingTools/TrajectoryState"/>
 <use   name="rootrflx"/>
+<Flags CXXFLAGS="-g -D=EDM_ML_DEBUG"/>
```
needs to be removed

slava77 commented on TrackPropagation/Geant4e/src/Geant4ePropagator.cc:
```diff
@@ -0,0 +1,588 @@
+	return TsosPP(TrajectoryStateOnSurface(tParsDest, curvError, pDest, side),
+			finalPathLength);
+}
+/*
```
are these commented out parts needed?
if not, please cleanup

lviliani committed:
Update BuildFile.xml

lviliani committed:
Update Geant4ePropagator.cc

lviliani commented:
@civanch We haven't modified the part about sensitive detectors in GeometryProducer.h.

cmsbuild commented:
Pull request #7505 was updated. @civanch, @nclopezo, @cvuosalo, @mdhildreth, @cmsbuild, @slava77 can you please check and sign again.

civanch commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented on TrackingTools/PatternTools/src/TrajectoryExtrapolatorToLine.cc:
```diff
@@ -13,7 +13,8 @@ TrajectoryStateOnSurface TrajectoryExtrapolatorToLine::extrapolate(const FreeTra
 
-  FreeTrajectoryState fastFts(fts.parameters());
+//  FreeTrajectoryState fastFts(fts.parameters());
+  FreeTrajectoryState fastFts = fts;  
```
can this trigger changes in tracks?

VinInn replied:
Lorenzo, could you please justify the modification?

VinInn replied:
-1
This is clearly a computational performance regression, need to be properly quantified.

slava77 commented:
I was expecting no changes in RECO of the tracks from this PR.

Looking at 25202 (ttbar pu35@25ns) I see about 10% increase in CPU spent in TrackProducer (details coming) and also a noticeable increase in duplicate tracks:
![wf25202_duprate_vs_eta](https://cloud.githubusercontent.com/assets/4676718/6022025/a0689ed8-ab7f-11e4-82f8-e49ecd82cb3c.png)

@rovere @VinInn 
Marco, Vincenzo, what do you think?

slava77 commented:
Here are a few more plots with noticeable differences
(plain, non-MC truth, plots show noticeable changes in conversionStepTracks in several tested samples) 

This is run1 ttbar with PU : this diff looks beyond just numerical variation
![all_sign510vsorig_ttbarpuwf202p0c_recotracks_conversionsteptracks__reco_obj_dxy](https://cloud.githubusercontent.com/assets/4676718/6022298/6747d6da-ab81-11e4-9213-dc2bd513aa10.png)

Note that in this sample the duplicate rate is essentially unchanged.
So, maybe the result in 25202 is some accidental unstable sequence disturbed by a numerical variation.
(all samples that I looked at suggest that there is at least a numerical precision level variation)

slava77 commented:
uhm, igprof tells me the TrackProducer cost is about the same (within 1%), in disagreement with the timer.

VinInn commented on RecoTracker/TrackProducer/interface/TrackProducerAlgorithm.h:
```diff
@@ -136,6 +138,7 @@ class TrackProducerAlgorithm {
   reco::TrackBase::TrackAlgorithm algo_;
   bool reMatchSplitHits_;
   bool geometricInnerState_;
+  bool usePropagatorForPCA_;
```
where is initialized?
please add = false;

slava77 replied:
this may actually explain what I see: after rerunning with igprof I get different results compared to without it

VinInn commented on RecoTracker/TrackProducer/interface/TrackProducerAlgorithm.h:
```diff
@@ -60,6 +60,8 @@ class TrackProducerAlgorithm {
 	  conf_.getParameter<bool>( "GeometricInnerState" ) : true);
```
three bools, three different ways to initialize them!

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7505/2297/summary.html

slava77 commented:
now that jenkins got around to making the comparisons, there are similar differences as what I've posted in jenkins tests as well

https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_4_X_2015-02-02-0200+7505/5863/validateJR/all_OldVSNew_RunPhoton2012Bwf4p53/all_OldVSNew_RunPhoton2012Bwf4p53c_recoTracks_conversionStepTracks__reRECO_obj_dxy.png

lenzip commented:
Hi,
     ok thanks,
we are looking into it.
Cheers
Giulio
On 03/02/2015 16:53, Slava Krutelyov wrote:
>
> now that jenkins got around to making the comparisons, there are 
> similar differences as what I've posted in jenkins tests as well
>
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_4_X_2015-02-02-0200+7505/5863/validateJR/all_OldVSNew_RunPhoton2012Bwf4p53/all_OldVSNew_RunPhoton2012Bwf4p53c_recoTracks_conversionStepTracks__reRECO_obj_dxy.png
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/cms-sw/cmssw/pull/7505#issuecomment-72674930>.
>

lviliani committed:
Update TrackProducerAlgorithm.h

Fixed uninitialized variable

lviliani committed:
Update TrajectoryExtrapolatorToLine.cc

Fixed initialization of FreeTrajectoryState

lviliani commented:
Dear All,
we have fixed the missing initialization in TrackProducerAlgoritm.h and we have fixed the assignment in TrajectoryExtrapolatorToLine.cc.
As far as we can see TrajectoryExtrapolatorToLine is not used in the standard tracking and anywhere in cmssw.
We have run runTheMatrix for sample 25202 with the latest commits and we don't see any differences.

cmsbuild commented:
Pull request #7505 was updated. @civanch, @nclopezo, @cvuosalo, @mdhildreth, @cmsbuild, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

civanch commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 69a4cc8b27f392247ba1083b0e5410da22e5225b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7505/2320/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7505/2320/summary.html

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/7505	 69a4cc8
after the fix the differences went away.
the timing is the same.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_4_X IBs unless changes (tests are also fine). This pull request will be automatically merged.
